images/shared-images-controller/main.go

```diff
@@ -0,0 +1,155 @@
+		log.WithError(err).Errorf("Failed to get latest kubevirtci tag")
+		return "", err
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		log.WithError(err).Errorf("Reading latest kubevirtci tag failed")
+		return "", err
+	}
+	tag = string(body)
+	return strings.TrimSpace(tag), nil
+}
+
+func pullRequiredImages(ctx context.Context, tag string) error {
+	var versions []string
+
+	endpoint := "https://api.github.com/"
+	client, err := github.NewEnterpriseClient(endpoint, endpoint, nil)
+	if err != nil {
+		log.WithError(err)
+		return err
+	}
+	context := context.Background()
+	releases, _, err := client.Repositories.ListReleases(context, "kubernetes", "kubernetes", nil)
+	if err != nil {
+		log.WithError(err).Errorf("Failed to list releases from kubernetes/kubernetes")
+		return err
+	}
+	minors := querier.LastThreeMinor(uint(1), releases)
+	for _, minor := range minors {
+		r := querier.ParseRelease(minor)
+		releaseVer := fmt.Sprintf("%s.%s", r.Major, r.Minor)
+		versions = append(versions, releaseVer)
+	}
+
+	log.Infoln("Last three minor releases", versions)
+
+	imageNames := map[string]struct{}{}
+	kubevirtci_repo := "quay.io/kubevirtci/"
+
+	imageList, err := images.List(ctx, nil)
+	if err != nil {
+		log.WithError(err).Errorf("Failed to list images")
+		return err
+	}
+	for _, i := range imageList {
+		for _, imageName := range i.RepoTags {
+			imageNames[imageName] = struct{}{}
+		}
+	}
+
+	for _, version := range versions {
+		log.Infoln("Kubevirt Provider version: ", version)
+		name := fmt.Sprintf("%sk8s-%s:%s", kubevirtci_repo, version, tag)
+		if _, exists := imageNames[name]; exists {
+			log.Infoln("Image already present:", name)
+			continue
+		}
+		log.Infoln("Pulling image: ", name)
+		_, err := images.Pull(ctx, name, nil)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func cleanOldImages(ctx context.Context, tag string) error {
+	var deleteImages []string
+	imageList, err := images.List(ctx, nil)
+	if err != nil {
+		log.WithError(err).Errorf("Failed to list images")
+		return err
+	}
+	for _, i := range imageList {
+		for _, repoTag := range i.RepoTags {
+			if strings.Contains(repoTag, tag) {
+				log.Infof("%s is a required image", repoTag)
+				continue
+			}
+			log.Infoln("Deleting image:", repoTag)
+			deleteImages = append(deleteImages, repoTag)
+
+		}
+	}
+	log.Infoln("images to delete:", deleteImages)
+	if len(deleteImages) > 0 {
+		_, err := images.Remove(ctx, deleteImages, nil)
+		if err != nil {
+			return err[0]
+		}
+	}
+	return nil
+}
+
+func main() {
+
+	var period time.Duration = 6
+
+	socket := "unix:/run/podman/podman.sock"
+
+	if os.Getenv("XDG_RUNTIME_DIR") != "" {
+		sock_dir := os.Getenv("XDG_RUNTIME_DIR")
+		socket = "unix:" + sock_dir + "/podman/podman.sock"
+	}
+	connText, err := bindings.NewConnection(context.Background(), socket)
+	if err != nil {
+		log.WithError(err).Fatalf("Could not connect to podman socket %d", socket)
+	}
+	for {
```

Comment:
I hope I am not overthinking this, but what if the process gets killed in the middle of an action like deleting images here? Do we need to care about broken image downloads? I guess not, since podman takes care of that? But what about if the deletion is only half way done? Does podman take care of that also?

However, I've been thinking about a more elegant way of running a periodic process, and I've found [Golang Ticker](https://pkg.go.dev/time#Ticker).

I don't think we need to do this now, but maybe in a later step. WDYT?

Reply:
> I hope I am not overthinking this, but what if the process gets killed in the middle of an action like deleting images here? Do we need to care about broken image downloads? I guess not, since podman takes care of that? But what about if the deletion is only half way done? Does podman take care of that also?

We would be relying on podman here - podman will try to download the image again when the loop comes around again so it should be ok on the next iteration. We would have to rely on podman for the image deletion too. 
> 
> However, I've been thinking about a more elegant way of running a periodic process, and I've found [Golang Ticker](https://pkg.go.dev/time#Ticker).
> 
> I don't think we need to do this now, but maybe in a later step. WDYT?

That does look good - we could definitely look into that later.